xdsl/pattern_rewriter.py

```diff
@@ -786,68 +785,128 @@ class PatternRewriteWalker:
+
+    def _process_worklist(self, listener: PatternRewriterListener) -> bool:
         """
-        Rewrite an operation, along with its regions.
-        Returns the next operation to iterate over.
+        Process the worklist until it is empty.
+        Returns true if any modification was done.
         """
-        # First, we rewrite the regions if needed
-        if self.walk_regions_first:
-            self._rewrite_op_regions(op)
+        rewriter_has_done_action = False
 
-        prev_op = op.prev_op
-        next_op = op.next_op
+        # Handle empty worklist
+        op = self._worklist.pop()
+        if op is None:
+            return rewriter_has_done_action
 
-        # We then match for a pattern in the current operation
+        # Create a rewriter on the first operation
         rewriter = PatternRewriter(op)
         rewriter.extend_from_listener(self.listener)
-        self.pattern.match_and_rewrite(op, rewriter)
-
-        if rewriter.has_done_action:
-            # If we produce new operations, we rewrite them recursively if requested
-            if self.apply_recursively:
-                if self.walk_reverse:
-                    for op in rewriter.iter_affected_ops_reversed():
-                        # return last affected op
-                        return op
-                    else:
-                        return prev_op
-                else:
-                    for op in rewriter.iter_affected_ops():
-                        # return first affected op
-                        return op
-                    else:
-                        return next_op
-
-            # Else, we rewrite only their regions if they are supposed to be
-            # rewritten after
-            else:
-                if not self.walk_regions_first:
-                    for new_op in rewriter.added_operations_before:
-                        self._rewrite_op_regions(new_op)
-                    if not rewriter.has_erased_matched_operation:
-                        self._rewrite_op_regions(op)
-                    for new_op in rewriter.added_operations_after:
-                        self._rewrite_op_regions(new_op)
-                return prev_op if self.walk_reverse else next_op
-
-        # Otherwise, we only rewrite the regions of the operation if needed
-        if not self.walk_regions_first:
-            self._rewrite_op_regions(op)
-        return prev_op if self.walk_reverse else next_op
-
-    def _rewrite_op_regions(self, op: Operation):
-        """
-        Rewrite the regions of an operation, and update the operation with the
-        new regions.
-        """
-        for region in op.regions:
-            blocks = reversed(region.blocks) if self.walk_reverse else region.blocks
-            for block in blocks:
-                iter_op = block.last_op if self.walk_reverse else block.first_op
-                while iter_op is not None:
-                    iter_op = self._rewrite_op(iter_op)
+
+        while True:
```

Comment:
Here the `while True` feels a little off also, since it's clearly one loop per operation, maybe it would be worth refactoring to something like this?
```suggestion
        while (op := self._worklist.pop()) is not None:
```

Reply:
Can't do really much more here sadly.
The reason why we cannot have this check at the beginning, is because we need to start with an `op` already, because the `PatternRewriter` should start with an `op` (I don't want to allow ill-formed `PatternRewriter`).
Main issue here to be honest, is that there are no do/while loop.
I changed it to make the structure more obvious, thoughts?

Reply:
I guess we could just hoist the first iteration of the loop and the repeat the loop body, but that does not sound a lot cleaner.

The other issue that makes this impossible is that we need to create the rewriter in the first place. So you could probably also implement something that really only happens in the first iteration. like:

```
rewriter = None
while (op := self._worklist.pop()) is not None:
  if not rewriter:
    rewriter = PatternRewriter(op)
    rewriter.extend_from_listener(listener)
```

But that is even worse, IMO. I don't have a very strong opinion on this. I think this and hoisting the first iteration are both fine solutions.

Reply:
Nope, because we need an `op` to create the `rewriter`.
To be fair, I feel that the current version is fine enough, with the current comments.